addons/resource/models/resource.py

```diff
@@ -536,193 +481,176 @@ def _schedule_days(self, cr, uid, id, days, day_date=None, compute_leaves=False,
-            default.update(name=_('%s (copy)') % (self.browse(cr, uid, id, context=context).name))
-        return super(resource_resource, self).copy(cr, uid, id, default, context)
-
+            default.update(name=_('%s (copy)') % (self.name))
+        return super(ResourceResource, self).copy(default)
+        
 
-class resource_calendar_leaves(osv.osv):
+class ResourceCalendarLeaves(models.Model):
     _name = "resource.calendar.leaves"
     _description = "Leave Detail"
-    _columns = {
-        'name' : fields.char("Name"),
-        'company_id' : fields.related('calendar_id','company_id',type='many2one',relation='res.company',string="Company", store=True, readonly=True),
-        'calendar_id' : fields.many2one("resource.calendar", "Working Time"),
-        'date_from' : fields.datetime('Start Date', required=True),
-        'date_to' : fields.datetime('End Date', required=True),
-        'resource_id' : fields.many2one("resource.resource", "Resource", help="If empty, this is a generic holiday for the company. If a resource is set, the holiday/leave is only for this resource"),
-    }
-
-    def check_dates(self, cr, uid, ids, context=None):
-        for leave in self.browse(cr, uid, ids, context=context):
-            if leave.date_from and leave.date_to and leave.date_from > leave.date_to:
-                return False
-        return True
-
-    _constraints = [
-        (check_dates, 'Error! leave start-date must be lower then leave end-date.', ['date_from', 'date_to'])
-    ]
-
-    def onchange_resource(self, cr, uid, ids, resource, context=None):
+
+    name = fields.Char()
+    company_id = fields.Many2one(string="Company", related='calendar_id.company_id', store=True, readonly=True)
+    calendar_id = fields.Many2one("resource.calendar", string="Working Time")
+    date_from = fields.Datetime(string='Start Date', required=True)
+    date_to = fields.Datetime(string='End Date', required=True)
+    resource_id = fields.Many2one("resource.resource", string="Resource",
+                                  help="If empty, this is a generic holiday for the company. If a resource is set, the holiday/leave is only for this resource")
+
+    @api.constrains('date_from', 'date_to')
+    def check_dates(self):
+        for leave in self:
+            if leave.date_from > leave.date_to:
+                raise ValidationError(_('Error! leave start-date must be lower then leave end-date.'))
+
+    @api.onchange('resource_id')
+    def onchange_resource(self):
         result = {}
-        if resource:
-            resource_pool = self.pool.get('resource.resource')
-            result['calendar_id'] = resource_pool.browse(cr, uid, resource, context=context).calendar_id.id
-            return {'value': result}
-        return {'value': {'calendar_id': []}}
+        if self.resource_id:
+            self.calendar_id = self.env['resource.resource'].browse(
+                self.resource_id).calendar_id.id
```

Comment:
no need to browse. self.resource_id itself recordset(s). 
directly use like this way : 
self.calendar_id = self.resource_id.calendar_id